File: general/make_readonly.c
#define _GNU_SOURCE
#include <sys/types.h>
#include <sys/mount.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <stdbool.h>
#include <signal.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stddef.h>
#include <dirent.h>
#include <stdint.h>
#include <stdio.h>
#include <errno.h>
#include <sched.h>

#define err_exit(msg)    do { perror(msg); exit(EXIT_FAILURE); \
                        } while (0)

// vimcmd: let &makeprg="gcc -Wall -W -g % -o %:r"
// vimcmd: !sudo chown root:root %:r
// vimcmd: !sudo chmod u+s %:r
//
// Silly little program to change a directory to read only for a shell session.
// The basic use case is for when you're poking around a machine that's quite
// important, and you're just feeling a little insecure about the pitfalls of
// unix ... "what if I get too comfortable and accidentaly type something
// stupid?".
// Have this program on your machine, run it to make the directory read-only,
// if you want to change something, then press ^D and change something before
// running this program again.
// Example:
//  ./make_readonly /home/hardenedapple /mnt/unused
// makes /home/hardenedapple  read-only
//
// Options that don't require a C program
// What I do here, but as a set of bash commands.
//  sudo unshare -m bash
//  mount --make-private /
//  mount -t overlay overlay -olowerdir=/home/hardenedapple:/mnt/usb /home/hardenedapple
//  cd ../hardenedapple
// Chroot into a new mount point.
//  sudo mount -t overlay overlay -olowerdir=/:<some empty directory> <mount point>;
//  sudo --preserve-env chroot --userspec=<user>:<group> <mount point>
//      # and to clean up
//  sudo umount <mount point>
//      // or
//  sudo mount -o rbind,ro / <mount point>
//  sudo mount --make-rprivate <mount point> # needed because systemd changed the defaults from private to shared
//  sudo --preserve-env chroot --userspec=<user>:<group> <mount point>
//      # and to clean up
//  sudo umount --recursive <mount point>
// Completely different approach, don't have to do anything special.
//  Create a new user on the machine
//  Give this user read-only access rights to everything
//  su - readonlyuser


void usage(char* command_name)
{
    printf("This command must be ran with two arguments.\n"
            "It starts a new shell session that can not modify files on the filesystem.\n"
            "Both arguments must be directories.\n"
            "The second directory given on the command line must be empty.\n"
            "Due to technical limitations, you can't use the root filesystem as the first argument.\n"
            "Usage: %s <re-mount point> <empty directory>\n", command_name);
    exit(1);
}

/*
 * Current understanding of the way things work:
 *  In order to make it so that mounts in one process don't affect mounts
 *  in another one you need to:
 *      Put them in different mount namespaces
 *      mark the mount point upon which you want to mount other things as
 *      private.
 *      then mount something on that private mount point.
 *
 *      Mark the mount point upon which you want to mount other things as
 *      private.
 *      Then mount something on that private mount point.
 *
 *      You then have to get into that mounted overlay.
 *      This can be done by a simple cd ../$(basename `pwd`)
 */

/*
 * Observations I don"t like / don"t understand / think might provide
 * useful information when trying to figure out why things aren"t working
 * the way I hope:
 *      vimcmd: r! uname --kernel-release | sed 's/^/                /'
            4.13.12-1-ARCH
 *      Man page on mount(2) says (in second paragraph of NOTES) that
 *      MS_MGC_VAL is no longer required and ignored if specified since
 *      Linux 2.4.
 *      But if I don"t include that flag I get EINVAL.
 *
 *      ls -l /proc/self/ns
 *      shows that when in the shell this command leaves you in, you have a
 *      different mount namespace to the one you started in.
 *
 *      What determines a peer group between namespaces?
 *      How would one change peer groups?
 *
 *      mount --make-private in the shell provided by this program shows
 *      that it removes the `shared:X` in the relevant line of
 *      /proc/self/mountinfo.
 *      But it doesn't remove the mount from other namespaces.
 *      In processes from other mount namespaces, the relevant line in
 *      /proc/self/mountinfo stays the same, with the `shared:X` still
 *      there.
 *      A process in one namespace changing files in the overlay fs
 *      propagates those changes to the other namespace, and unmounting the
 *      mount point in one namespace unmounts it to the other.
 */

// Documentation on overlayfs
// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/overlayfs.txt
// Make the root mount point private.

// Option 1
//  Clone to create a new namespace.
//  child -> Lose root priviledges.
//  child -> execv('bash')
//  parent -> wait()
int mount_readonly(
        uid_t orig_uid,
        const char * const mount_point,
        const char * const empty_dir)
{
    const char * const options_format = "lowerdir=%s:%s";
    // Don't actually need the + 1 here, but it saves anyone reading it the
    // thought "are we missing the NULL character ... oh no, it"s counted in
    // the `%s` characters of the format string.
    const size_t num_chars = strlen(mount_point) + strlen(empty_dir) + strlen(options_format) + 1;
    char* const options_string = malloc(num_chars);
    snprintf(options_string, num_chars, options_format, mount_point, empty_dir);
    if (unshare(CLONE_NEWNS) == -1) { err_exit("unshare"); }
    // Make entire system privately mounted.
    if (mount("none", "/", NULL, MS_REC|MS_PRIVATE, NULL)) { err_exit("mount make private"); }
    if (mount("overlay",
           mount_point,
           "overlay",
           MS_MGC_VAL,
           // Can"t *just* mount an overlayfs with only one filesystem.
           // Hence I"m using a filesystem I happen to know to be empty so that
           // things aren"t confusing.
           options_string)) {
        err_exit("mount overlay");
    }

    // Have to change directory to the mount point so that we enter the
    // overlayfs instead of staying in the original filesystem (which is not
    // mounted read-only).
    if (chdir(mount_point)) { err_exit("chdir"); }
    if (seteuid(orig_uid)) { err_exit("uid failed"); }

    char * const argv[] = { "/bin/bash", NULL };
    // environ is defined in <unistd.h> if _GNU_SOURCE is defined.
    execve("/bin/bash", argv, environ);
    err_exit("execve returned");
    return 0;
}

bool is_directory(const char * const name)
{
    struct stat file_info;
    if (stat(name, &file_info)) { err_exit("failure in stat"); }
    bool answer = S_ISDIR(file_info.st_mode);
    return answer;
}

bool is_empty_directory(const char * const name)
{
    if (!is_directory(name)) { return false; }
    DIR * dirstream = opendir(name);
    if (dirstream == NULL) { err_exit("opendir"); }
    errno = 0;
    uint8_t num_elements = 0;
    struct dirent *cur_dirent;
    while (num_elements < 3 && (cur_dirent = readdir(dirstream))) { 
        num_elements += 1;
    }
    if (errno != 0) { err_exit("readdir"); }
    if (closedir(dirstream)) { err_exit("closedir"); }
    return num_elements == 2;
}

int main(int argc, char *argv[])
{
    // Unfortunately, because of the need to change directory out of the
    // current mount and into the new mount, overlay mounting the "/" directory
    // doesn"t work.
    // This because we can"t move out of the "/" directory to be able to move
    // back into it.
    //
    // Because of this we need the user to specify the directory to overlay
    // with itself.
    if (argc != 3) { usage(argv[0]); }
    if (!is_directory(argv[1])) { usage(argv[0]); }
    if (!is_empty_directory(argv[2])) { usage(argv[0]); }
    if (!strcmp("/", argv[1])) { usage(argv[0]); }
    uid_t real = getuid();
    if (seteuid(0)) { err_exit("seteuid to root failed"); }
    mount_readonly(real, argv[1], argv[2]);

    return 0;
}

File: general/make_readonly.rs
#![feature(slice_patterns)]
extern crate libc;
extern crate errno;

use std::os::unix::process::CommandExt;

/*
 * N.b. has to be compiled with access to the libc and errno crates.
 * TODO???
 *  Maybe I should invoke whichever shell the user is currently running.
 *  Not something I actually want to do, but it makes sense as a functionality.
 */


#[derive(Debug)]
enum ArgumentIsNot {
    Given,
    Directory,
    Empty,
    Usable(::std::io::Error),
    Allowed,
}

// Rather than manually implement Debug to print out the enum type as well as the enum value, just
// do this.
#[derive(Debug)]
enum SillyFormattingTrick { ArgumentIsNot(ArgumentIsNot) }

fn usage(arg_problem: ArgumentIsNot) {
    let command_name = ::std::env::args().next().unwrap_or(String::from("\"Unknown Name\""));
    println!("Usage: {} <empty directory>", command_name);
    print!("Bind mounts the root filesystem read only in the given directory,");
    println!("then chroots into that directory.");
    println!("To avoid losing access to files, the argument must be an empty directory.");
    println!("Problem: {:?}", SillyFormattingTrick::ArgumentIsNot(arg_problem));
}

/// Parse the arguments so that
fn parse_args() -> Result<String, ArgumentIsNot> {
    let mut args = ::std::env::args();
    // Seeing as we have to iterate over the first argument here, we may as well check that the
    // first argument exists, and return error if it doesn't.
    // This first argument should be the name of the program.
    if args.next().is_none() { return Result::Err(ArgumentIsNot::Given) }
    match args.next() {
        None => return Result::Err(ArgumentIsNot::Given),
        Some(dirname) => {
            // We check for directory here to give nice error messages to the user.
            // We would find a problem if the given name was a file when attempting to read it as a
            // directory below, but then we would have to go backwards from the error message to
            // figure out what the problem was.
            match ::std::fs::metadata(&dirname) {
                Err(real_err) => return Result::Err(ArgumentIsNot::Usable(real_err)),
                Ok(info) => {
                    if !info.is_dir() { return Result::Err(ArgumentIsNot::Directory); }
                }
            }
            match ::std::fs::read_dir(&dirname) {
                Err(real_err) => return Result::Err(ArgumentIsNot::Usable(real_err)),
                Ok(mut dir_iter) => {
                    // It appears that the "." and ".." entries are ignored.
                    if !dir_iter.next().is_none() {
                        return Result::Err(ArgumentIsNot::Empty);
                    }
                }
            }
            if args.next().is_some() {
                Result::Err(ArgumentIsNot::Allowed)
            } else {
                Result::Ok(dirname)
            }
        }
    }
}

// If anything fails, the program exits, and because all state changes were done in a namespace
// isolating those changes everything is fine.
// Hence, no cleanup.
fn mount_with(mountpoint: String) -> String {
    let pathbuf = match std::env::current_dir() {
        Ok(x) => x,
        Err(ioerr) => return format!("{}", ioerr),
    };
    
    // Keep the CString values around so they don't get deallocated until the end of the block.
    let rootcstr = std::ffi::CString::new("/")
                            .expect("Failed to create CString from \"/\"");
    let mountcstr = std::ffi::CString::new(mountpoint)
                            .expect("Failed to create CString from given argument!!");
    let root = rootcstr.as_ptr();
    let mount = mountcstr.as_ptr();
    let unused = root;

    let orig_euid = unsafe { libc::getuid() };

    if unsafe { libc::unshare(libc::CLONE_NEWNS) } != 0 {
        return String::from("unshare error");
    }
    if unsafe { libc::mount(unused,
                            root,
                            std::ptr::null(),
                            libc::MS_REC | libc::MS_PRIVATE,
                            std::ptr::null()) } != 0 {
        return String::from("mount make private");
    }
    // n.b. running `strace mount -o rbind,ro / /mnt/usb` shows that program calls mount twice.
    // mount("/", "/mnt/usb", 0x55f6bee59480, MS_MGC_VAL|MS_RDONLY|MS_BIND|MS_REC, NULL) = 0
    // mount("none", "/mnt/usb", NULL, MS_RDONLY|MS_REMOUNT|MS_BIND|MS_REC, NULL) = 0
    //
    // While it looks to me like the first one should mount the bound mount read only (there is the
    // MS_RDONLY flag there after all) it doesn't.
    // Hence we have to make the second call too.
    if unsafe { libc::mount(root,
                            mount,
                            std::ptr::null(),
                            libc::MS_MGC_VAL | libc::MS_RDONLY | libc::MS_BIND | libc::MS_REC,
                            std::ptr::null()) } != 0 {
        return String::from("mount bind");
    }
    if unsafe { libc::mount(unused,
                            mount,
                            std::ptr::null(),
                            libc::MS_RDONLY | libc::MS_REMOUNT | libc::MS_BIND | libc::MS_REC,
                            std::ptr::null()) } != 0 {
        return String::from("mount make read only");
    }

    if unsafe { libc::chroot(mount) } != 0 { return String::from("chroot"); }
    match std::env::set_current_dir(pathbuf) {
        Ok(_) => {},
        Err(ioerr) => return format!("Error changing directory! {}", ioerr),
    }
 
    if unsafe { libc::seteuid(orig_euid) } != 0 { return String::from("seteuid back to user"); }

    format!("exec error {}", std::process::Command::new("/bin/bash").exec())
}

pub fn main() {
   match parse_args() {
        Ok(dirname) => {
            // If mount_with() returns, that's an error.
            eprintln!("Error during mount_with: {}", mount_with(dirname));
            eprintln!("{}", errno::errno());
        },
        Err(arg_problem) =>  usage(arg_problem)
    }
    ::std::process::exit(libc::EXIT_FAILURE);
}
